### PR TITLE
fix: encoding version type

### LINKED ISSUE #2124: JsonAbi encoding version type too strict
The [`encoding`](https://github.com/FuelLabs/fuels-ts/blob/master/packages/abi-coder/src/types/JsonAbi.ts) field type for the `JsonAbi` is too strict and preventing users from using a raw JSON with the `ContractFactory`. We should relax this type to allow users to pass a JSON object to the ContractFactory.

```console
@fuel-bridge/test-utils:build:     Types of property 'encoding' are incompatible.
@fuel-bridge/test-utils:build:       Type 'string' is not assignable to type 'EncodingVersion'.
@fuel-bridge/test-utils:build: src/utils/fuels/getOrDeployFuelTokenContract.ts(56,7): error TS2345: Argument of type '{ encoding: string; types: { typeId: number; type: string; components: { name: string; type: number; typeArguments: any; }[]; typeParameters: number[]; }[]; functions: { inputs: { name: string; type: number; typeArguments: any; }[]; name: string; output: { ...; }; attributes: { ...; }[]; }[]; loggedTypes: { ...; }[]...' is not assignable to parameter of type 'JsonAbi | Interface<JsonAbi>'.
```
[Source](https://github.com/FuelLabs/fuel-bridge/blob/main/packages/test-utils/src/utils/fuels/getOrDeployFuelTokenContract.ts#L56)

### PR DESCRIPTION
**Pull request by petertonysmith94**
Closes #2124 

---

- Changed the `version` field on the `JsonAbi` to a string.
- Added validation on the encoding field.

### CODE REVIEW

File: packages/abi-coder/src/utils/json-abi.ts [hunk 1]
```diff
@@ -3,6 +3,39 @@ import { ErrorCode, FuelError } from '@fuel-ts/errors';
...
+    case ENCODING_V0:
+      return ENCODING_V0;
+
+    default:
+      throw new FuelError(
+        ErrorCode.UNSUPPORTED_ENCODING_VERSION,
+        `Encoding version '${encoding}' is unsupported.`
+      );
+  }
+};
```
Comment (nedsalk) [inline packages/abi-coder/src/utils/json-abi.ts:37]:
```suggestion
export const getEncodingVersion = (encoding?: string): EncodingVersion => {
  switch (encoding) {
    case undefined:
    case ENCODING_V0:
        return ENCODING_V0;

    case ENCODING_V1:
      return ENCODING_V1;

    default:
      throw new FuelError(
        ErrorCode.UNSUPPORTED_ENCODING_VERSION,
        `Encoding version '${encoding}' is unsupported.`
      );
  }
};
```
And you can remove the DEFAULT_ENCODING_VERSION constant with this.

File: packages/abi-coder/src/types/JsonAbi.ts [hunk 2]
```diff
@@ -2,15 +2,12 @@
...
-
-import type { EncodingVersion } from '../utils/constants';
-
 export interface JsonAbi {
   readonly types: readonly JsonAbiType[];
   readonly loggedTypes: readonly JsonAbiLoggedType[];
   readonly functions: readonly JsonAbiFunction[];
   readonly configurables: readonly JsonAbiConfigurable[];
-  readonly encoding?: EncodingVersion;
+  readonly encoding?: string;
```
Comment (nedsalk) [inline packages/abi-coder/src/types/JsonAbi.ts:10]: Can you explain why we'd go from `EncodingVersion` to `string` when this field is optional to begin with so the user doesn't have to type it at all?

File: packages/abi-coder/src/utils/constants.ts [hunk 3]
```diff
@@ -25,9 +25,11 @@ export const genericRegEx = /^generic (?<name>\w+)$/;
 /**
  * Encoding versions
  */
-export type EncodingVersion = '0' | '1';
 export const ENCODING_V0 = '0';
 export const ENCODING_V1 = '1';
+export const DEFAULT_ENCODING_VERSION = ENCODING_V0;
```
Comment (maschad) [inline packages/abi-coder/src/utils/constants.ts:30]: Is the default always going to be `v0` ? If so then I don't think we need this constant and then @nedsalk 's [comment applies](https://github.com/FuelLabs/fuels-ts/pull/2126/files#r1574608969)

**Review by maschad — CHANGES_REQUESTED** [review]
Nice, but also need to address the comments that you made as well.

**Review by maschad — APPROVED** [review]
Great work!

**Review by nedsalk — APPROVED** [review]

**Review by danielbate — APPROVED** [review]
Good find 🦊 

This PR introduces some duplication for encoding version retrieval, however approving as this resolves the bug and I think that can be cleaned up #2109

**Pull request merged**